### PR TITLE
Add qchem gates and generators 

### PR DESCRIPTION
**Pull request by maliasadi**
**Context:**
This PR adds all unsupported parameterised gates and generators for adjoint-diff including,
- SingleExcitation,
- SingleExcitationPlus,
- SingleExcitationMinus,
- DoubleExcitation,
- DoubleExcitationPlus,
- DoubleExcitationMinus,

**Benchmark:**

### CODE REVIEW

**Comment by chaeyeunpark** [comment]
Thanks, @maliasadi for the great works! But can we merge this after my PR (#274) as it involves several changes in kernels?

**Comment by maliasadi** [comment]
> Thanks, @maliasadi for the great works! But can we merge this after my PR (#274) as it involves several changes in kernels?

Sure thing @chaeyeunpark!

**Comment by chaeyeunpark** [comment]
Hi @maliasadi, I will review this PR when the changes in `master` are merged.

File: pennylane_lightning/src/gates/cpu_kernels/QChemGateImplementations.hpp [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+/**
+ * @file
+ * Defines qchem kernel functions with generators
+ */
+#pragma once
+
+#include "cpu_kernels/GateImplementationsLM.hpp"
+#include "cpu_kernels/GateImplementationsPI.hpp"
+
+namespace Pennylane::Gates {
```
Comment (chaeyeunpark) [inline pennylane_lightning/src/gates/cpu_kernels/QChemGateImplementations.hpp:23]: Is there a specific reason to separate these functions in another file?

File: pennylane_lightning/src/simulator/DefaultKernels.hpp [hunk 2]
```diff
@@ -87,6 +97,18 @@ namespace Pennylane {
...
+    std::pair{Gates::GeneratorOperation::SingleExcitationMinus,
+              Gates::KernelType::LM},
+    std::pair{Gates::GeneratorOperation::SingleExcitationPlus,
+              Gates::KernelType::LM},
+    std::pair{Gates::GeneratorOperation::DoubleExcitation,
+              Gates::KernelType::PI},
+    std::pair{Gates::GeneratorOperation::DoubleExcitationMinus,
+              Gates::KernelType::PI},
+    std::pair{Gates::GeneratorOperation::DoubleExcitationPlus,
+              Gates::KernelType::PI},
```
Comment (chaeyeunpark) [inline pennylane_lightning/src/simulator/DefaultKernels.hpp:111]: :+1:

File: pennylane_lightning/src/tests/Test_DynamicDispatcher.cpp [hunk 3]
```diff
@@ -83,7 +83,7 @@ constexpr void testAllGatesForKernelIter(RandomEngine &re,
     if constexpr (idx < static_cast<int>(GateOperation::END)) {
         constexpr auto gate_op = static_cast<GateOperation>(idx);
 
-        for (size_t num_qubits = 3; num_qubits <= max_num_qubits;
+        for (size_t num_qubits = 4; num_qubits <= max_num_qubits;
```
Comment (chaeyeunpark) [inline pennylane_lightning/src/tests/Test_DynamicDispatcher.cpp:86]: Is it possible to change the starting number to the number of qubits that the gate requires i.e. `Util::static_lookup<gate_op>(Constant::gate_wires)`?

**Review by chaeyeunpark — COMMENTED** [review]
Thanks, @maliasadi for amazing work! I just have a few questions.

File: pennylane_lightning/src/tests/Test_DynamicDispatcher.cpp (hunk 3 above)
Comment (maliasadi) [reply pennylane_lightning/src/tests/Test_DynamicDispatcher.cpp:86]: Fixed it! Thanks :slightly_smiling_face:

File: pennylane_lightning/src/gates/cpu_kernels/GateImplementationsLM.hpp [hunk 4]
```diff
@@ -98,6 +98,9 @@ class GateImplementationsLM : public PauliGenerator<GateImplementationsLM> {
         GateOperation::IsingXX,
         GateOperation::IsingYY,
         GateOperation::IsingZZ,
+        GateOperation::SingleExcitation,
+        GateOperation::SingleExcitationMinus,
+        GateOperation::SingleExcitationPlus,
```
Comment (mlxd) [inline pennylane_lightning/src/gates/cpu_kernels/GateImplementationsLM.hpp:103]: Do we need the double excitation ops here too? Or is there a single implementation of these?

File: pennylane_lightning/src/gates/cpu_kernels/GateImplementationsLM.hpp [hunk 5]
```diff
@@ -112,6 +115,9 @@ class GateImplementationsLM : public PauliGenerator<GateImplementationsLM> {
         GeneratorOperation::IsingYY,
         GeneratorOperation::IsingZZ,
         GeneratorOperation::ControlledPhaseShift,
+        GeneratorOperation::SingleExcitation,
+        GeneratorOperation::SingleExcitationMinus,
+        GeneratorOperation::SingleExcitationPlus,
```
Comment (mlxd) [inline pennylane_lightning/src/gates/cpu_kernels/GateImplementationsLM.hpp:120]: And here?

**Review by mlxd — APPROVED** [review]
Awesome work @maliasadi 
Nothing major to add, just a single question.

File: pennylane_lightning/src/gates/cpu_kernels/GateImplementationsLM.hpp (hunk 4 above)
Comment (maliasadi) [reply pennylane_lightning/src/gates/cpu_kernels/GateImplementationsLM.hpp:103]: This is due to this fact that less-memory (LS) implementation of ops with more than 2-qubit  hasn't explored in lightning and I'm also not sure even with exploring them we will gain significant performance improvement. For ops like Toffoli, CSWAP and DoubleExcitation*, we rely on precomputed-indices (PI) implementation.

File: pennylane_lightning/src/gates/cpu_kernels/GateImplementationsLM.hpp (hunk 5 above)
Comment (maliasadi) [reply pennylane_lightning/src/gates/cpu_kernels/GateImplementationsLM.hpp:120]: The same as above.

File: pennylane_lightning/src/gates/cpu_kernels/GateImplementationsLM.hpp (hunk 4 above)
Comment (mlxd) [reply pennylane_lightning/src/gates/cpu_kernels/GateImplementationsLM.hpp:103]: Sounds good, thanks Ali

**Review by chaeyeunpark — APPROVED** [review]
Thanks, @maliasadi again for a great job! I just have one more question.

File: pennylane_lightning/src/tests/Test_DynamicDispatcher.cpp [hunk 6]
```diff
@@ -82,12 +82,16 @@ constexpr void testAllGatesForKernelIter(RandomEngine &re,
...
+        constexpr auto num_wires =
+            std::max(static_cast<size_t>(
+                         Util::static_lookup<gate_op>(Constant::gate_wires)),
+                     min_num_wires);
+        for (size_t num_qubits = num_wires; num_qubits <= max_num_qubits;
              num_qubits++) {
             testDynamicDispatch<PrecisionT, ParamT, GateImplementation,
                                 gate_op>(re, num_qubits);
-        }
+        };
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply pennylane_lightning/src/tests/Test_DynamicDispatcher.cpp:94]: Thank you @chaeyeunpark! Good catch :smile:

**Review by isschoch — APPROVED** [review]
Looks good. nice & clean code!

**Pull request merged**